libethash-cuda/CUDAMiner.cpp

```diff
@@ -454,65 +454,66 @@ void CUDAMiner::search(
         {
             cudaStream_t stream = m_streams[current_index];
             buffer = m_search_buf[current_index];
-            // Wait for stream batch to complete
+
+            // Wait for stream batch to complete and immediately
+            // store number of processed hashes
             CUDA_SAFE_CALL(cudaStreamSynchronize(stream));
+            addHashCount(batch_size);
+
             if (shouldStop())
             {
                 m_new_work.store(false, std::memory_order_relaxed);
                 done = true;
                 stop = true;
             }
+
             // See if we got solutions in this batch
-            uint32_t found_count = buffer->count;
+            uint32_t found_count = std::min((unsigned)buffer->count, SEARCH_RESULTS);
             if (found_count)
             {
+
                 buffer->count = 0;
-                uint64_t nonces[SEARCH_RESULTS];
-                h256 mixes[SEARCH_RESULTS];
-                // handle the highly unlikely possibility that there are more
-                // solutions found than we can handle
-                if (found_count > SEARCH_RESULTS)
-                    found_count = SEARCH_RESULTS;
                 uint64_t nonce_base = current_nonce - streams_batch_size;
-                // stash the solutions, so we can reuse the search buffer
-                for (unsigned int j = 0; j < found_count; j++)
-                {
-                    nonces[j] = nonce_base + buffer->result[j].gid;
-                    if (s_noeval)
-                        memcpy(mixes[j].data(), (void*)&buffer->result[j].mix,
-                            sizeof(buffer->result[j].mix));
-                }
-                // restart the stream on the next batch of nonces
-                if (!done)
-                    run_ethash_search(
-                        s_gridSize, s_blockSize, stream, buffer, current_nonce, m_parallelHash);
-                // Pass the solutions up to the higher level
+                
+                // Pass the solution(s) for submission
+                uint64_t minerNonce;
+                h256 minerMix;
```

Comment:
This declaration could be moved into the memcpy block. It is only used there.

Reply:
I'm always concerned about the right place where to locate var initialization.
As I carry my heritage from .Net, var initialization within loops is never a good practice.
Does C++ handle this differently ? (Do not know much about how the assembler works)

Reply:
In C++ this is fine.

Reply:
Nice

Reply:
There is no extra overhead in C++. Stack offsets for local storage are calculated at compile time. For simple numeric variables  there is no run-time initialization anyway, for others like strings or more complex objects, that would need to be initialized, you would need to initialize them manually with an assignment statement before using it in any logic that requires it to be preset to a known value.

Limiting the scope of a variable is actually a recommended practice in C++ for the following reasons:

- It prevents the possible misuse of the variable outside of its intended context.
- Most important, it allows the reader of the code to easily recognize that the variable is only used within that scope and is not part of any other logic.